core/chains/evm/gas/universal_estimator.go

```diff
@@ -0,0 +1,412 @@
+package gas
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	bigmath "github.com/smartcontractkit/chainlink-common/pkg/utils/big_math"
+
+	commonfee "github.com/smartcontractkit/chainlink/v2/common/fee"
+	feetypes "github.com/smartcontractkit/chainlink/v2/common/fee/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/gas/rollups"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+)
+
+// metrics are thread safe
+var (
+	promUniversalEstimatorGasPrice = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "gas_price_updater",
+		Help: "Sets latest gas price (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+	promUniversalEstimatorBaseFee = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "base_fee_updater",
+		Help: "Sets latest BaseFee (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+	promUniversalEstimatorMaxPriorityFeePerGas = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "max_priority_fee_per_gas_updater",
+		Help: "Sets latest MaxPriorityFeePerGas (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+	promUniversalEstimatorMaxFeePerGas = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "max_fee_per_gas_updater",
+		Help: "Sets latest MaxFeePerGas (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+)
+
+const (
+	MinimumBumpPercentage   = 10 // based on geth's spec
+	ConnectivityPercentile  = 80
+	BaseFeeBufferPercentage = 40
+)
+
+type UniversalEstimatorConfig struct {
+	BumpPercent  uint16
+	CacheTimeout time.Duration
+	EIP1559      bool
+
+	BlockHistorySize uint64
+	RewardPercentile float64
+	HasMempool       bool
```

Comment:
I think this was discussed offline but `HasMempool` may be a misleading name. ZkSync has a first come first serve model but also has a mempool. Maybe we could call this `IsFirstComeFirstServe`?

Reply:
I don't think what was mentioned offline was accurate, and if I'm not mistaken Friedemann verified it. Either you have a mempool and transactions are prioritized based on priorityFee or you don't have a mempool and priorityFee is just a placeholder to comply with EIP-1559.

Reply:
Maybe it's a future change? I see this [blog post](https://blog.matter-labs.io/zksync-2-0-update-dynamic-fees-milestone-completed-c6620a3d2618) by Matter Labs about EIP-1559 support being different on zkSync. Regardless my concern was just that the config naming doesn't really reflect what it's controlling. It's inferring a behavior based on some property of the chain. I feel like we should go with something more self-descriptive.

Reply:
To me, having a mempool and processing transactions in an FCFS manner is tautological, but the term `mempool` is broadly used when we want to describe this property, so it seemed like the better fit. In any case, I don't have a strong preference, but I want us to use a name that can be easily recognizable by people without much context.

Reply:
Tbh my initial suggestion is also not really self-descriptive of what the config controls. Maybe `IgnorePriorityFees` is more in line with the functionality. But I'm ok keeping it the way it is. I'll leave this comment unresolved for now in case other reviewers have opinions

Reply:
In the EVM world, the notion of having a mempool is equal to prioritizing transactions based on fees. zkSync doesn't have a mempool since transaction fees are ignored.
I believe the best name for this should be whatever is more recognizable by our engineers, so it might make sense to ask them.